kong/core/handler.lua

```diff
@@ -119,9 +121,25 @@ return {
       end
 
       -- if set `host_header` is the original header to be preserved
-      var.upstream_host = host_header or 
+      var.upstream_host = host_header or
           balancer_address.hostname..":"..balancer_address.port
 
+      -- X-Forwarded Headers
+      local realip_remote_addr = var.realip_remote_addr
+
+      if not singletons.ip.trusted(realip_remote_addr) then
+        var.upstream_x_forwarded_proto = var.scheme
+        var.upstream_x_forwarded_host  = var.host
+        var.upstream_x_forwarded_port  = var.server_port
+      end
+
+      local http_x_forwarded_for = var.http_x_forwarded_for
+      if http_x_forwarded_for then
+        var.upstream_x_forwarded_for = http_x_forwarded_for .. ", " .. realip_remote_addr
+
+      else
+        var.upstream_x_forwarded_for = var.remote_addr
```

Comment:
I think some comments here would be very beneficial to avoid future confusion, regarding the usage of `realip_remote_addr` vs `var.remove_addr`, and why we need to build the `X-Forwarded-For` upstream header ourselves (due to the ngx_http_proxy_module limitation we discovered).